keeper/component/dispatcher/state.go

```diff
@@ -0,0 +1,504 @@
+				ccID, err := core.ParseCrossChainID(pk.K3())
+				if err != nil {
+					return collections.Triple[int32, string, string]{}, errors.Wrap(
+						err,
+						"error parsing destination cross-chain ID",
+					)
+				}
+
+				return collections.Join3(
+					int32(ccID.GetProtocolId()),
+					ccID.GetCounterpartyId(),
+					pk.K4(),
+				), nil
+			},
+		),
+	}
+}
+
+// ====================================================================================================
+// Dispatched Amount
+// ====================================================================================================
+
+func (d *Dispatcher) GetDispatchedAmount(
+	ctx context.Context,
+	sourceID *core.CrossChainID,
+	destID *core.CrossChainID,
+	denom string,
+) *dispatchertypes.DispatchedAmountEntry {
+	key := collections.Join4(
+		int32(sourceID.GetProtocolId()),
+		sourceID.GetCounterpartyId(),
+		destID.ID(),
+		denom,
+	)
+
+	amount, err := d.dispatchedAmounts.Get(ctx, key)
+	if err != nil {
+		d.logger.Error("received an error getting dispatched amounts",
+			"error", err,
+			"source_id", sourceID.ID(),
+			"dest_id", destID.ID(),
+			"denom", denom,
+		)
+		amount = dispatchertypes.AmountDispatched{
+			Incoming: math.ZeroInt(),
+			Outgoing: math.ZeroInt(),
+		}
+	}
+
+	return &dispatchertypes.DispatchedAmountEntry{
+		SourceId:         sourceID,
+		DestinationId:    destID,
+		Denom:            denom,
+		AmountDispatched: amount,
+	}
+}
+
+func (d *Dispatcher) HasDispatchedAmount(
+	ctx context.Context,
+	sourceID *core.CrossChainID,
+	destID *core.CrossChainID,
+	denom string,
+) bool {
+	da := d.GetDispatchedAmount(ctx, sourceID, destID, denom)
+
+	return da.AmountDispatched.IsPositive()
+}
+
+func (d *Dispatcher) SetDispatchedAmount(
+	ctx context.Context,
+	sourceID *core.CrossChainID,
+	destID *core.CrossChainID,
+	denom string,
+	amountDispatched dispatchertypes.AmountDispatched,
+) error {
+	key := collections.Join4(
+		int32(sourceID.GetProtocolId()),
+		sourceID.GetCounterpartyId(),
+		destID.ID(),
+		denom,
+	)
+
+	return d.dispatchedAmounts.Set(ctx, key, amountDispatched)
+}
+
+func (d *Dispatcher) GetDispatchedAmountsBySourceProtocolID(
+	ctx context.Context,
+	protocolID core.ProtocolID,
+) []*dispatchertypes.DispatchedAmountEntry {
+	amounts := []*dispatchertypes.DispatchedAmountEntry{}
+
+	rng := collections.NewPrefixedQuadRange[int32, string, string, string](int32(protocolID))
+
+	err := d.dispatchedAmounts.Walk(
+		ctx,
+		rng,
+		func(k DispatchedAmountsKey, v dispatchertypes.AmountDispatched) (stop bool, err error) {
+			entry, err := d.GetDispatchedAmountEntryFromKey(ctx, k)
+			if err != nil {
+				return true, err
+			}
+
+			amounts = append(amounts, &entry)
+
+			return false, nil
+		},
+	)
+	if err != nil {
+		d.logger.Error(
+			"error in dispatched amounts walking by source protocol ID",
+			"error",
+			err,
+		)
```

Comment:
I'm not entirely sure about this logs in the queries honestly. We always return the query result even if the record doesn't exist, and  we log it as an error. 
Especially for node runners/infra could  be confusing

<img width="1436" height="190" alt="Image" src="https://github.com/user-attachments/assets/13d5e656-7e8c-444c-8b61-16fca1b6da8c" />